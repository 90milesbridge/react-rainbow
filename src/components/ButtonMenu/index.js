import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import ButtonIcon from './../ButtonIcon';
import MenuContent from './menuContent';
import { Provider } from './context';
import {
    findItemByKey,
    findItemIndex,
    insertChildOrderly,
    getChildMenuItemNodes,
} from './utils';
import {
    UP_KEY,
    DOWN_KEY,
    ESCAPE_KEY,
    TAB_KEY,
    ENTER_KEY,
} from './../../constants';

export default class ButtonMenu extends Component {
    constructor(props) {
        super(props);
        this.containerRef = React.createRef();
        this.buttonRef = React.createRef();
        this.toggleMenu = this.toggleMenu.bind(this);
        this.handleKeyPressed = this.handleKeyPressed.bind(this);
        this.handleKeyDownPressed = this.handleKeyDownPressed.bind(this);
        this.handleKeyUpPressed = this.handleKeyUpPressed.bind(this);
        this.handleKeyEscapePressed = this.handleKeyEscapePressed.bind(this);
        this.handleKeyEnterPressed = this.handleKeyEnterPressed.bind(this);
        this.handleClick = this.handleClick.bind(this);
        this.closeMenu = this.closeMenu.bind(this);

        this.registerChild = this.registerChild.bind(this);
        this.unregisterChild = this.unregisterChild.bind(this);

        this.state = {
            isOpen: false,
            childrenRefs: [],
            childFocusedIndex: -1,
            context: {
                privateOnClose: this.closeMenu,
                privateRegisterChild: this.registerChild,
                privateUnregisterChild: this.unregisterChild,
            },
        };
    }

    getContainerClassNames() {
        const { isOpen } = this.state;
        const { className } = this.props;

        return classnames('slds-dropdown-trigger slds-dropdown-trigger_click', {
            'slds-is-open': isOpen,
        }, className);
    }

    getDropdownClassNames() {
        const { menuAlignment, menuSize, isLoading } = this.props;
        return classnames(
            'slds-dropdown',
            `slds-dropdown_${menuAlignment}`,
            `slds-dropdown_${menuSize}`,
            { 'slds-icon_large': isLoading },
        );
    }

    focusChild(index) {
        const { childrenRefs } = this.state;
        this.setState({ childFocusedIndex: index });
        return childrenRefs[index].focus();
    }

    handleKeyDownPressed() {
        const { childrenRefs, childFocusedIndex } = this.state;
        const lastChild = childrenRefs.length - 1;
        const isLastChild = childFocusedIndex === lastChild;
        const isBetweenFirstAndLast = childFocusedIndex >= 0 && !isLastChild;
        const isInvalidIndexOrLast = childFocusedIndex === -1 || isLastChild;

        if (isBetweenFirstAndLast) {
            return this.focusChild(childFocusedIndex + 1);
        }
        if (isInvalidIndexOrLast) {
            return this.focusChild(0);
        }
        return null;
    }

    handleKeyUpPressed() {
        const { childrenRefs, childFocusedIndex } = this.state;
        const isFirstChild = childFocusedIndex === 0;
        const isValidIndexGreaterThanFirst = childFocusedIndex >= 0 && !isFirstChild;
        const isInvalidIndexOrFirst = childFocusedIndex === -1 || isFirstChild;
        const lastChild = childrenRefs.length - 1;

        if (isValidIndexGreaterThanFirst) {
            return this.focusChild(childFocusedIndex - 1);
        }
        if (isInvalidIndexOrFirst) {
            return this.focusChild(lastChild);
        }
        return null;
    }

    handleKeyEscapePressed() {
        this.toggleMenu();
        return this.buttonRef.current.focus();
    }

    handleKeyEnterPressed() {
        const { childrenRefs, childFocusedIndex } = this.state;
        const isValidIndex = childFocusedIndex >= 0;
        if (isValidIndex) {
            return childrenRefs[childFocusedIndex].click();
        }
        return null;
    }

    focusMatchedItem(matchedItem) {
        const { childrenRefs } = this.state;
        const itemIndex = findItemIndex(childrenRefs, matchedItem);
        return this.focusChild(itemIndex);
    }

    findItemByKeyPressed(key) {
        const { childrenRefs, childFocusedIndex, matchedKeyPressed } = this.state;
        if (matchedKeyPressed) {
            const newChildrenRefs = childrenRefs.slice(childFocusedIndex + 1);
            const matchedItem = findItemByKey(key, newChildrenRefs);
            if (matchedItem) {
                return this.focusMatchedItem(matchedItem);
            }
            const newMatchedItem = findItemByKey(key, childrenRefs);
            if (newMatchedItem) this.focusChild(findItemIndex(childrenRefs, newMatchedItem));
            return null;
        }
        const matchedItem = findItemByKey(key, childrenRefs);
        if (matchedItem) {
            this.setState({ matchedKeyPressed: key });
            return this.focusMatchedItem(matchedItem);
        }
        return null;
    }

    handleKeyPressed(event) {
        const { isOpen } = this.state;
        if (isOpen) {
            if (event.keyCode !== TAB_KEY) event.preventDefault();
            const keyHandlerMap = {
                [DOWN_KEY]: this.handleKeyDownPressed,
                [UP_KEY]: this.handleKeyUpPressed,
                [ESCAPE_KEY]: this.handleKeyEscapePressed,
                [TAB_KEY]: this.toggleMenu,
                [ENTER_KEY]: this.handleKeyEnterPressed,
            };
            if (keyHandlerMap[event.keyCode]) {
                return keyHandlerMap[event.keyCode]();
            }
            return this.findItemByKeyPressed(event.key);
        }
        return null;
    }

    registerChild(childRef) {
        const { childrenRefs } = this.state;
        const [...nodes] = getChildMenuItemNodes(this.containerRef.current);
        const newChildrenRefs = insertChildOrderly(childrenRefs, childRef, nodes);
        this.setState({
            childrenRefs: newChildrenRefs,
        });
    }

    unregisterChild(childRef) {
        const { childrenRefs } = this.state;
        const newChildrenRefs = childrenRefs.filter(child => child !== childRef);
        this.setState({
            childrenRefs: newChildrenRefs,
        });
    }

    handleClick(event) {
        const isClickInsideButtonMenu = this.containerRef.current.contains(event.target);
        if (isClickInsideButtonMenu) {
            return null;
        }
        return this.closeMenu();
    }

    openMenu() {
        window.addEventListener('click', this.handleClick);
        return this.setState({
            isOpen: true,
        });
    }

    closeMenu() {
        window.removeEventListener('click', this.handleClick);
        return this.setState({
            isOpen: false,
            childFocusedIndex: -1,
        });
    }

    toggleMenu() {
        const { isOpen } = this.state;
        if (isOpen) {
            return this.closeMenu();
        }
        return this.openMenu();
    }

    render() {
        const {
            style,
            title,
            assistiveText,
            iconName,
            buttonVariant,
            buttonSize,
            disabled,
            tabIndex,
            onFocus,
            onBlur,
            isLoading,
            children,
        } = this.props;
        const { context } = this.state;
        const ariaLabel = title || assistiveText;

        return (
            <div
                role="presentation"
                className={this.getContainerClassNames()}
                style={style}
                onKeyDown={this.handleKeyPressed}
                ref={this.containerRef}>

                <ButtonIcon
                    disabled={disabled}
                    tabIndex={tabIndex}
                    variant={buttonVariant}
                    size={buttonSize}
                    ariaHaspopup
                    title={title}
                    assistiveText={assistiveText}
                    iconName={iconName}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    onClick={this.toggleMenu}
                    ref={this.buttonRef} />

                <div className={this.getDropdownClassNames()}>
                    <ul className="slds-dropdown__list" role="menu" aria-label={ariaLabel}>
                        <MenuContent isLoading={isLoading}>
                            <Provider value={context}>
                                {children}
                            </Provider>
                        </MenuContent>
                    </ul>
                </div>
            </div>
        );
    }
}

ButtonMenu.propTypes = {
    className: PropTypes.string,
    style: PropTypes.object,
    title: PropTypes.string,
    assistiveText: PropTypes.string,
    iconName: PropTypes.string,
    menuAlignment: PropTypes.oneOf([
        'left', 'right', 'bottom',
    ]),
    menuSize: PropTypes.oneOf([
        'xx-small',
        'x-small',
        'small',
        'medium',
        'large',
    ]),
    /** The variant for the button with predefined styles. */
    buttonVariant: PropTypes.oneOf([
        'base',
        'border-filled',
        'border-inverse',
        'border',
        'brand',
        'inverse',
        'container',
    ]),
    /** The button icon size. */
    buttonSize: PropTypes.oneOf([
        'xx-small',
        'x-small',
        'small',
        'medium',
        'large',
    ]),
    disabled: PropTypes.bool,
    isLoading: PropTypes.bool,
    children: PropTypes.node,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    tabIndex: PropTypes.number,
};

ButtonMenu.defaultProps = {
    className: '',
    style: {},
    title: undefined,
    assistiveText: undefined,
    iconName: 'utility:down',
    menuAlignment: 'left',
    menuSize: 'xx-small',
    buttonVariant: 'border-filled',
    buttonSize: 'medium',
    disabled: false,
    isLoading: false,
    children: null,
    onFocus: () => {},
    onBlur: () => {},
    tabIndex: undefined,
};
