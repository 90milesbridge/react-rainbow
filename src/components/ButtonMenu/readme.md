button menu base

    <div className="slds-p-vertical_large slds-align_absolute-center" >
        <ButtonGroup>
            <ButtonIcon iconName="utility:add" variant="border-filled" />
            <ButtonIcon iconName="utility:edit" variant="border-filled" />
            <ButtonIcon iconName="utility:paste" variant="border-filled" />
            <ButtonMenu menuAlignment="right" menuSize="x-small">
                <MenuItem label="Menu Item One" />
                <MenuItem label="Menu Item Two" />
                <MenuItem label="Menu Item Three" />
            </ButtonMenu>
        </ButtonGroup>
    </div>


button menu with subheaders

    <div className="slds-p-vertical_large slds-align_absolute-center" >
        <ButtonGroup>
            <ButtonIcon iconName="utility:add" variant="border-filled" />
            <ButtonIcon iconName="utility:edit" variant="border-filled" />
            <ButtonIcon iconName="utility:paste" variant="border-filled" />
            <ButtonMenu menuAlignment="right" menuSize="x-small">
                <MenuItem label="Menu header" variant="header" />
                <MenuItem label="Menu Item One" />
                <MenuItem label="Menu Item Two" />
                <MenuItem label="Menu Item Three" />
                <MenuItem label="Menu header" variant="header" />
                <MenuItem label="Menu Item One" />
                <MenuItem label="Menu Item Two" />
            </ButtonMenu>
        </ButtonGroup>
    </div>


button menu with icons

    <div className="slds-p-around_large slds-align_absolute-center" >
        <div className="slds-m-horizontal_xx-large">
            <ButtonMenu menuAlignment="right" menuSize="x-small">
                <MenuItem 
                    label="Right Icon"
                    iconName="utility:table" 
                    iconPosition="right" />

                <MenuItem 
                    label="Right Icon"
                    iconName="utility:kanban" 
                    iconPosition="right" />

                <MenuItem 
                    label="Right Icon"
                    iconName="utility:side_list" 
                    iconPosition="right" />

            </ButtonMenu>
        </div>
        <div className="slds-m-horizontal_xx-large">
            <ButtonMenu menuSize="x-small">
               <MenuItem 
                   label="Left Icon"
                   iconName="utility:announcement" 
                   iconPosition="left" />

               <MenuItem 
                   label="Left Icon"
                   iconName="utility:announcement" 
                   iconPosition="left" />

               <MenuItem 
                   label="Left Icon"
                   iconName="utility:announcement" 
                   iconPosition="left" />

            </ButtonMenu>
        </div>
    </div>


button menu position variants

    <div className="slds-p-around_large slds-align_absolute-center" >
        <div className="slds-m-horizontal_xx-large">
            <ButtonMenu menuAlignment="right" menuSize="x-small">
                <MenuItem label="Menu Positioned Right" />
                <MenuItem label="Menu Positioned Right" />
                <MenuItem label="Menu Positioned Right" />
            </ButtonMenu>
        </div>
        <div className="slds-m-horizontal_xx-large">
            <ButtonMenu menuAlignment="center" menuSize="x-small">
                <MenuItem label="Menu Positioned Center" />
                <MenuItem label="Menu Positioned Center" />
                <MenuItem label="Menu Positioned Center" />
            </ButtonMenu>
        </div>
        <div className="slds-m-horizontal_xx-large">
            <ButtonMenu menuAlignment="left" menuSize="x-small">
                <MenuItem label="Menu Positioned Left" />
                <MenuItem label="Menu Positioned Left" />
                <MenuItem label="Menu Positioned Left" />
            </ButtonMenu>
        </div>
    </div>


button menu width variants

    <div className="slds-m-horizontal_xx-large slds-m-vertical_large" >
        <div className="slds-grid slds-grid_align-spread">
            <ButtonMenu menuSize="xx-small">
                <MenuItem label="xx-small" />
                <MenuItem label="xx-small" />
                <MenuItem label="xx-small" />
            </ButtonMenu>
            <ButtonMenu menuSize="x-small" menuAlignment="right">
                <MenuItem label="x-small" />
                <MenuItem label="x-small" />
                <MenuItem label="x-small" />
            </ButtonMenu>
            <ButtonMenu menuSize="small" menuAlignment="right">
                <MenuItem label="small" />
                <MenuItem label="small" />
                <MenuItem label="small" />
            </ButtonMenu>
        </div>
        <div className="slds-p-top_xx-large">
            <ButtonMenu menuSize="medium">
                <MenuItem label="Menu medium" />
                <MenuItem label="Menu medium" />
                <MenuItem label="Menu medium" />
            </ButtonMenu>
        </div>
        <div className="slds-p-top_xx-large">
            <ButtonMenu menuSize="large">
                <MenuItem label="Menu large" />
                <MenuItem label="Menu large" />
                <MenuItem label="Menu large" />
            </ButtonMenu>
        </div>
    </div>
    

button menu loading

    <div className="slds-p-vertical_large slds-align_absolute-center" >
        <ButtonGroup>
            <ButtonIcon iconName="utility:add" variant="border-filled" />
            <ButtonIcon iconName="utility:edit" variant="border-filled" />
            <ButtonIcon iconName="utility:paste" variant="border-filled" />
            <ButtonMenu menuAlignment="right" menuSize="x-small" isLoading />
        </ButtonGroup>
    </div>