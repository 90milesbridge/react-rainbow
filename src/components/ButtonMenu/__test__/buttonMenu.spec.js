import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ButtonMenu from './../';

describe('<ButtonMenu/>', () => {
    it('should focus the button when the focus method is called', () => {
        const component = mount(
            <ButtonMenu />,
        );

        component.instance().focus();
        const focusedElementDataId = document.activeElement.getAttribute('data-id');
        const buttonDataId = component.find('button').prop('data-id');
        expect(focusedElementDataId).toBe(buttonDataId);
    });
    it('should blur the button when the blur method is called', () => {
        const component = mount(
            <ButtonMenu />,
        );
        const instance = component.instance();
        const buttonDataId = component.find('button').prop('data-id');

        instance.focus();
        expect(document.activeElement.getAttribute('data-id')).toBe(buttonDataId);
        instance.blur();
        expect(document.activeElement.getAttribute('data-id')).toBeNull();
    });
    it('should set the title passed in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu title="my title">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('title')).toBe('my title');
    });
    it('should set the assistiveText passed in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu assistiveText="assistive description">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('assistiveText')).toBe('assistive description');
    });
    it('should set the iconName passed in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu iconName="utility:user">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('iconName')).toBe('utility:user');
    });
    it('should set the buttonVariant passed as variant in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu buttonVariant="brand">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('variant')).toBe('brand');
    });
    it('should set the buttonSize passed as size in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu buttonSize="large">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('size')).toBe('large');
    });
    it('should set the disabled passed in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu disabled>
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('disabled')).toBe(true);
    });
    it('should set the tabIndex passed in the ButtonIcon', () => {
        const component = mount(
            <ButtonMenu tabIndex={0}>
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ButtonIcon').prop('tabIndex')).toBe(0);
    });
    it('should set the isLoading passed in the MenuContent', () => {
        const component = mount(
            <ButtonMenu isLoading>
                <span />
            </ButtonMenu>,
        );
        expect(component.find('MenuContent').prop('isLoading')).toBe(true);
    });
    it('should render the children passed', () => {
        const component = mount(
            <ButtonMenu>
                <span data-id="menu-children">menu content</span>
            </ButtonMenu>,
        );
        expect(component.find('span[data-id="menu-children"]').text()).toBe('menu content');
    });
    it('should set the role as menu in the ul element', () => {
        const component = mount(
            <ButtonMenu>
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ul').prop('role')).toBe('menu');
    });
    it('should set the title passed as the aria-label in the ul element', () => {
        const component = mount(
            <ButtonMenu title="button title" assistiveText="assistive description">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ul').prop('aria-label')).toBe('button title');
    });
    it('should set the assistiveText passed as the aria-label in the ul element when the title is not passed', () => {
        const component = mount(
            <ButtonMenu assistiveText="assistive description">
                <span />
            </ButtonMenu>,
        );
        expect(component.find('ul').prop('aria-label')).toBe('assistive description');
    });
    it('should set the right class names to the container element when the menu is opened', () => {
        const component = mount(
            <ButtonMenu>
                <span />
            </ButtonMenu>,
        );
        component.find('ButtonIcon').simulate('click');
        expect(component.find('div[role="presentation"]').prop('className')).toBe('slds-dropdown-trigger slds-dropdown-trigger_click slds-is-open');
    });
    it('should have the right class names', () => {
        const component = renderer.create(
            <ButtonMenu>
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuAlignment is left and isLoading is true', () => {
        const component = renderer.create(
            <ButtonMenu menuAlignment="left" isLoading>
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuAlignment is right and menuSize is xx-small', () => {
        const component = renderer.create(
            <ButtonMenu menuAlignment="right" menuSize="xx-small">
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuSize is x-small', () => {
        const component = renderer.create(
            <ButtonMenu menuSize="x-small">
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuSize is small', () => {
        const component = renderer.create(
            <ButtonMenu menuSize="small">
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuSize is medium', () => {
        const component = renderer.create(
            <ButtonMenu menuSize="medium">
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when menuSize is large', () => {
        const component = renderer.create(
            <ButtonMenu menuSize="large">
                <span />
            </ButtonMenu>,
        );
        expect(component).toMatchSnapshot();
    });
});
