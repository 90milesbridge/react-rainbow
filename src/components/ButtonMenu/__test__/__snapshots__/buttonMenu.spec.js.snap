// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonMenu/> should have the right class names 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_xx-small"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuAlignment is left and isLoading is true 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_xx-small slds-icon_large"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <div
        className="slds-spinner slds-spinner_small"
        role="status"
      >
        <span
          className="slds-assistive-text"
        >
          loading menu
        </span>
        <div
          className="slds-spinner__dot-a"
        />
        <div
          className="slds-spinner__dot-b"
        />
      </div>
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuAlignment is right and menuSize is xx-small 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_right slds-dropdown_xx-small"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuSize is large 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_large"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuSize is medium 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_medium"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuSize is small 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_small"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;

exports[`<ButtonMenu/> should have the right class names when menuSize is x-small 1`] = `
<div
  className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
  onKeyDown={[Function]}
  role="presentation"
>
  <button
    aria-haspopup={true}
    className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled"
    data-id="button-icon-element"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon"
    >
      <use
        href="/icons/utility-sprite/svg/symbols.svg#down"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </button>
  <div
    className="slds-dropdown slds-dropdown_left slds-dropdown_x-small"
  >
    <ul
      className="slds-dropdown__list"
      role="menu"
    >
      <span />
    </ul>
  </div>
</div>
`;
